docs: Add how-to guide for panel plugin transformations (#2756)
grafana/plugin-tools: docusaurus/docs/how-to-guides/panel-plugins/work-with-transformations.md

:::tip

For a step-by-step guide to making your own panel plugin, refer to our [Tutorial for panel plugins](../../tutorials/build-a-panel-plugin.md).

:::

## How the data pipeline works

When a user queries data in Grafana, the results pass through a pipeline before reaching your panel plugin:

```
Data source query → Transformations → Field overrides → Panel plugin
```

1. **Data source query**: The data source returns one or more [data frames](../../key-concepts/data-frames).
2. **Transformations**: Grafana applies any user-configured transformations sequentially, in the order the user arranged them, modifying the structure or content of the data frames.
3. **Field overrides**: Grafana applies field configuration (units, thresholds, colors, display names) to the transformed data.
4. **Panel plugin**: Your plugin receives the final `PanelData` object via `props.data`.

> Your plugin always receives already-transformed data. You never need to apply transformations yourself, and there is no way for a panel to opt out of them or access the untransformed data.

Reviewer: ```suggestion
Your plugin always receives already-transformed data. **You never need to apply transformations yourself, and there is no way for a panel to opt out of them or access the untransformed data**.
```

I'd stress this!

Author: Thanks, fixed in https://github.com/grafana/plugin-tools/pull/2756/changes/78ef26fb3765509644448a25a1dffcf386ef5b35 👍

## Handle multiple data frames

Transformations can change the number of frames in either direction: **Merge** and **Join by field** combine multiple frames into one, while **Partition by values** splits a single frame into many. Always consider that `data.series` may contain any number of frames:

```ts title="src/components/MyPanel.tsx"
export function MyPanel({ data, options }: Props) {
  const frames = data.series;

  if (frames.length === 0 || frames.every((frame) => frame.length === 0)) {
    throw new Error('Query returned no data.');
  }

  // Option A: Work with all frames
  for (const frame of frames) {
    // Process each frame
  }

  // Option B: Let the user select which frame to display
  const frameIndex = Math.max(0, Math.min(options.frameIndex ?? 0, frames.length - 1));
  const frame = frames[frameIndex];
}
```

## Detect structure changes with `structureRev`

> If your panel performs expensive processing when the data shape changes — for example, rebuilding a chart configuration — use `data.structureRev` to tell structural changes apart from value-only updates. Grafana increments this revision counter whenever the structure of the data frames changes: fields added, removed, renamed, or retyped. Editing transformations typically changes the structure, so it also bumps the revision.

Reviewer: ```suggestion
If your panel performs expensive processing when the data shape changes such as rebuilding a chart configuration, use `data.structureRev` to tell structural changes apart from value-only updates. Grafana increments this revision counter whenever the structure of the data frames changes: fields added, removed, renamed, or retyped. Editing transformations typically changes the structure, so it also bumps the revision.
```

Author: Fixed in https://github.com/grafana/plugin-tools/pull/2756/changes/b520e0907f3549f2bed7e601b90c6b9dc1634ffd 👍

export function MyPanel({ data, options }: Props) {
  const frame = data.series[0];

  const valueField = frame.fields.find((field) =>
    options.valueFieldName ? field.name === options.valueFieldName : field.type === FieldType.number
  );
}
```

This approach works well with transformations because users can reshape the data first, then select the appropriate fields in your panel options.

## Handle multiple data frames

Transformations can change the number of frames in either direction: **Merge** and **Join by field** combine multiple frames into one, while **Partition by values** splits a single frame into many. Always consider that `data.series` may contain any number of frames:

```ts title="src/components/MyPanel.tsx"
export function MyPanel({ data, options }: Props) {
  const frames = data.series;

  if (frames.length === 0 || frames.every((frame) => frame.length === 0)) {
    throw new Error('Query returned no data.');
  }

  // Option A: Work with all frames
  for (const frame of frames) {
    // Process each frame
  }

  // Option B: Let the user select which frame to display
  const frameIndex = Math.min(options.frameIndex ?? 0, frames.length - 1);

>   const frame = frames[frameIndex];

Author: Fixed in https://github.com/grafana/plugin-tools/pull/2756/commits/2610991d270901e7172164274832215efd1d8837.
